src/qpx_mmc.cpp

```diff
@@ -0,0 +1,4244 @@
+    break;
+  case 0xE4:
+    strcpy(str, "PLEXTOR_AS_RD                                 ");
+    break;
+  case 0xE5:
+    strcpy(str, "PLEXTOR_AS_WR                                 ");
+    break;
+
+  case 0xE7:
+    strcpy(str, "SBC_FLUSH_CACHE                               ");
+    break;
+
+  case 0xE9:
+    strcpy(str, "PLEXTOR_MODE                                  ");
+    break;
+  case 0xEA:
+    strcpy(str, "PLEXTOR_QCHECK                                ");
+    break;
+  case 0xEB:
+    strcpy(str, "PLEXTOR_PREC_SPD                              ");
+    break;
+  case 0xED:
+    strcpy(str, "PLEXTOR_MODE2                                 ");
+    break;
+
+  case 0xEE:
+    strcpy(str, "PLEXTOR_RESTART                               ");
+    break;
+    //		case 0xEF: strcpy(str,"PLEXTOR     REBOOT ??? "); break;
+
+  case 0xF1:
+    strcpy(str, "PLEXTOR_EEPROM_READ                           ");
+    break;
+
+  case 0xF3:
+    strcpy(str, "PLEXTOR_SCAN_TA_FETE                          ");
+    break;
+  case 0xF5:
+    strcpy(str, "PLEXTOR_FETE_READOUT                          ");
+    break;
+  default:
+    strcpy(str, "*unknown*                                     ");
+  }
+  printf("[%02X]  %s", opcode, str);
+  return 0;
+}
+
+int probe_drive(const char *path, int idx) {
+  drive_info *drive;
+  int inq;
+  //	printf("Trying device: %s\n", path);
+  drive = new drive_info(path);
+  drive->silent++;
+  inq = inquiry(drive);
+  drive->silent--;
+  switch (inq) {
+  case 0x00:
+#if 0
+//				strcpy(drvtbl[drvcnt],_devtbl[i]);
+        printf("[%02d] %s %s: %s %s %s\n",(drive->ven_ID & vendor_mask) ? "*" : " ",
+               drvcnt, drive->device, drive->ven, drive->dev, drive->fw);
+#else
+    printf("D: [%02d] '%s': '%s' '%s' '%s'\n", idx, drive->device, drive->ven,
+           drive->dev, drive->fw);
+#endif
+    break;
+  case ERR_NO_DEV:
+    //	if (!drive->silent) printf("%s: no device found\n",drive->device);
+    break;
+  case ERR_NO_SCSI:
+    //	if (!drive->silent) printf("%s: not a valid SCSI
+    //device\n",drive->device);
+    break;
+  case ERR_NO_MMC:
+    //	printf("%s: %s %s %s",drive->device,drive->ven,drive->dev,drive->fw);
+    //	printf(": device is not MMC compliant\n");
+    //	for (j=0; j<8; j++)
+    //	    printf("Ver ID [%d] = 0x%04X\n",j,drive->ver_id[j]);
+    break;
+  default:
+    //	printf("%s: ???\n",drive->device);
+    break;
+  }
+  delete drive;
+  return inq;
+}
+
+int scanbus(int vendor_mask) {
+  int i;
+  //	int inq;
+  int drvcnt = 0;
+
+#ifndef DEVTBL_DIRECT
+  DIR *dir;
+  struct dirent *dentry;
+  struct stat st;
+  str_dev devstr;
+  int dlen;
+#endif
+
+  printf("** scanning IDE/SATA/SCSI buses...\n");
+
+#ifdef DEVTBL_DIRECT // used in win32 only
+  for (i = 0; strlen(_devtbl[i].name) > 0; i++) {
+    if (!probe_drive(_devtbl[i].name, drvcnt))
+      drvcnt++;
+  }
+#else
+  for (i = 0; strlen(_devtbl[i].name) > 0; i++) {
+    dlen = strlen(_devtbl[i].name);
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)